src/snowflake/connector/file_transfer_agent.py

```diff
@@ -380,507 +390,269 @@ def execute(self):
+        num_total_files = len(metas)
+        transfer_metadata = TransferMetadata()  # this is protected by cv_chunk_process
+        is_upload = self._command_type == CMD_TYPE_UPLOAD
+        exception_caught_in_callback: Optional[Exception] = None
+
+        def notify_file_completed():
+            # Increment the number of completed files, then notify the main thread.
+            with cv_main_thread:
+                transfer_metadata.num_files_completed += 1
+                cv_main_thread.notify()
+
+        def preprocess_done_cb(
+            success: bool,
+            result: Any,
+            file_meta: SnowflakeFileMeta,
+            done_client: "SnowflakeStorageClient",
+        ):
+            if not success:
+                logger.debug(f"Failed to prepare {done_client.meta.name}.")
+                if is_upload:
+                    done_client.finish_upload()
+                else:
+                    done_client.finish_download()
+                notify_file_completed()
+            elif done_client.meta.result_status == ResultStatus.SKIPPED:
+                # this case applies to upload only
+                notify_file_completed()
+            else:
+                logger.debug(f"Finished preparing file {done_client.meta.name}")
+                with cv_chunk_process:
+                    while transfer_metadata.chunks_in_queue > 2 * max_concurrency:
+                        logger.debug(
+                            "Chunk queue busy, waiting in file done callback..."
+                        )
+                        cv_chunk_process.wait()
+                    for _chunk_id in range(done_client.num_of_chunks):
+                        _callback = partial(
+                            transfer_done_cb,
+                            done_client=done_client,
+                            chunk_id=_chunk_id,
+                        )
+                        if is_upload:
+                            network_tpe.submit(
+                                function_and_callback_wrapper,
+                                # Work fn
+                                done_client.upload_chunk,
+                                # Callback fn
+                                _callback,
+                                file_meta,
+                                # Arguments for work fn
+                                _chunk_id,
+                            )
+                        else:
+                            network_tpe.submit(
+                                function_and_callback_wrapper,
+                                # Work fn
+                                done_client.download_chunk,
+                                # Callback fn
+                                _callback,
+                                file_meta,
+                                # Arguments for work fn
+                                _chunk_id,
+                            )
+                        transfer_metadata.chunks_in_queue += 1
```

Comment:
once submit, chunks_in_queue adds one. A large file with many chunks(e.g. 1000 chunks) will cross the 2 * max_concurrency limit (e.g. 32) a lot, in this case, every transfer_done_cb will trigger  cv_chunk_process.notify() and cause the while loop (at line 444) of some other threads beging awaked and then wait again, maybe a semaphore could help reducing the cost here.

       while transfer_metadata.chunks_in_queue > 2 * max_concurrency:
                        cv_chunk_process.wait()"

Reply:
Good point, a semaphore is better fitted here!
The use-case is not something that we recommend though, do you think that we could treat this as an enhancement later instead of delaying this feature?